library/network/src/lib/network/firewalld.rb

```diff
@@ -0,0 +1,324 @@
+      begin
+        fwcmd = FWCmd.new("--permanent --service=#{service} --get-description")
+        fwcmd.output
+      rescue FirewallCMDError
+        "Unknown description for #{service}"
+      end
+    end
+
+    # True if FirewallD supports the service. Returns true/false.
+    def is_service_supported?(service)
+      supported_services = get_supported_services
+      supported_services.include?(service)
+    end
+
+    # True if service is enabled in zone. Returns true/false
+    def is_service_enabled?(zone, service)
+      do_cmd("--permanent --zone=#{zone} --query-service=#{service}")
+    end
+
+    # Return port/protocol pair as defined in the service file
+    def get_service_ports(service)
+      fwcmd = FWCmd.new("--permanent --service=#{service} --get-ports")
+      fwcmd.output.strip
+    end
+
+    # Return port/protocol pair as defined in the service file
+    def get_service_protocol(service)
+      fwcmd = FWCmd.new("--permanent --service=#{service} --get-protocol")
+      fwcmd.output.strip
+    end
+
+    # True if port is enabled in zone.
+    def is_port_enabled?(zone, port)
+      do_cmd("--permanent --zone=#{zone} --query-port=#{port}")
+    end
+
+    # True if protocol is enabled in zone.
+    def is_protocol_enabled?(zone, protocol)
+     do_cmd("--permanent --zone=#{zone} --query-protocol=#{protocol}")
+    end
+
+    # True to add service to zone.
+    def add_service_to_zone(zone, service)
+      do_cmd("--permanent --zone=#{zone} --add-service=#{service}")
+    end
+
+    # True to add port to zone.
+    def add_port_to_zone(zone, port)
+      do_cmd("--permanent --zone=#{zone} --add-port=#{port}")
+    end
+
+    # True to add port to zone.
+    def add_protocol_to_zone(zone, protocol)
+      do_cmd("--permanent --zone=#{zone} --add-protocol=#{protocol}")
+    end
+
+    # True to remove service from zone.
+    def remove_service_from_zone(zone, service)
+      do_cmd("--permanent --zone=#{zone} --remove-service=#{service}")
+    end
+
+    # True to remove service from zone.
+    def remove_port_from_zone(zone, port)
+      do_cmd("--permanent --zone=#{zone} --remove-port=#{port}")
+    end
+
+    # True to remove service from zone.
+    def remove_protocol_from_zone(zone, protocol)
+      do_cmd("--permanent --zone=#{zone} --remove-protocol=#{protocol}")
+    end
+
+    # True if masquerade is enabled in zone
+    def is_masquerade_enabled?(zone)
+      do_cmd("--permanent --zone=#{zone} --query-masquerade")
+    end
+
+    # True to enable masquerade to zone.
+    def add_masquerade(zone)
+      return true if is_masquerade_enabled?(zone)
+      do_cmd("--permanent --zone=#{zone} --add-masquerade")
+    end
+
+    # True to remove masquerade from zone.
+    def remove_masquerade(zone)
+      return true if not is_masquerade_enabled?(zone)
+      do_cmd("--permanent --zone=#{zone} --remove-masquerade")
+    end
+
+    ### Logging ##
+
+    # True if $value logging is enabled
+    def is_log_denied?(value)
```

Comment:
When I read "is log denied?" I understand it as "is logging disabled?" Only when I read the [man page](https://twoerner.fedorapeople.org/firewalld/doc/firewall-cmd.html) I find
> ... logging rules right before reject and drop rules in the INPUT, ...

so it means "are denied *packets* logged?" This is looks like poor naming in upstream. I would name it
```rb
# @return [String] all, unicast, broadcast, multicast or off. 
def log_denied_packets; ...; end
# @param kind [String] all, unicast, broadcast, multicast or off. 
def log_denied_packets=(kind); ...; end
```